### ISSUE TITLE
validation accuracy 99%,  predict() 70% on validation accuracy FOLDER   -> error?

### ISSUE THREAD

**Issue by ethex2020** [keras-team/keras#14099]
Hello,

i have either made a mistake or found a huge error respectively a conceptional error in keras/tensorflow.

 I want to know whether predict() makes the same predictions as the internal function in the training process which is output printed while training in the terminal. I am speaking of train/validation == acc/val_acc. 
I tried the model InceptionV3 on Cats vs Dogs and get as the best result:

Epoch 175/300
200/200 [==============================] - 45s 227ms/step - loss: 0.0134 - acc: 0.9960 - val_loss: 0.0011 - val_acc: 0.9925

The sets of images are stored in the folders train(4000 images)/validation(2000 images).
After the training was done, I took the validation folder and tested the trained network with it.
The validation folder given to the predict() method of keras. According to the training statistics of the validation accuracy, my expectations were that the results should be close to 99%. But the prediction for “cat” gave me 70% correct  classified and 30% misclassified results. For “dogs” I got also the same amount of true classified results of 70% and 30% incorrect classifications.

I tried to find errors in my approach like using different data-set. But the result was the same. I searched through the forums to find an answer, but none of the ideas worked.

Do you have an explanation for this strange behavior of the predict function?

I tried my algorithm with these development environments:
tensorflow 1.10.0
keras 2.2.2
cudatoolkit 9.2
cudnn 7.2.1

and

tensorflow 2.1
keras 2.3.1
cudatoolkit 10.1.243
cudnn 7.6.5 

I tried and the result were equal:

'binary'
predictions = Dense(1, activation='sigmoid')(x)
model.compile(loss='binary_crossentropy', optimizer=optimizers.RMSprop(lr=1e-5), metrics=['acc'])

'categorical_crossentropy' - metrics
predictions = Dense(2, activation='softmax')(x)
model.compile(loss='categorical_crossentropy', optimizer=optimizers.RMSprop(lr=1e-5), metrics=['categorical_accuracy'])

'categorical_crossentropy'
predictions = Dense(2, activation='softmax')(x)
model.compile(loss='categorical_crossentropy', optimizer=optimizers.RMSprop(lr=1e-5), metrics=['acc'])

# source code
https://github.com/ethex2020/predict70

Thank you!

**Comment by ZohebAbai** [comment]
Hi,
I don't think there is a problem with predict(), but let's figure it out.
Can you try the following? 

Use 
**path_best_model = 'F:\\petsTF20model_weights\\0.9925__175-0.00-0.9960_CatsVRDogs_4000_inceptionV3_softmaxP98_TF2.1win.h5'**
 instead of 'F:\\petsTF20model_weights\\CATSvsDOGS_4000_2000_inceptionV3_softmax_model.save98_TF2.1win.h5'
and use this in
model = load_model(path_best_model)
**and comment out
model.load_weights('F:\\petsTF20model_weights\\0.9925__175-0.00-0.9960_CatsVRDogs_4000_inceptionV3_softmaxP98_TF2.1win.h5')**

Let us know how this works out?

One side note: Instead of predict() you can try 
model.evaluate_generator(validation_generator, steps=validation_steps, verbose =1)

**Comment by ethex2020** [comment]
I made the changes to the code as you suggest. 

path_best_model = 'F:\petsTF20model_weights\0.9925__175-0.00-0.9960_CatsVRDogs_4000_inceptionV3_softmaxP98_TF2.1win.h5'
model = load_model(path_best_model)

The results are still the same. No changes.
https://github.com/ethex2020/predict70/blob/master/v2_search_for_error/CATSvsDOGS_2000_1000_predict_sortANDcopy_classify_loop--SOFTMAX_InceptionV3-Copy6_sent.ipynb

I try to figure it out if their is a easy way to get prediction results for one image with evaluate_generator methode.
I can make some changes so evaluate_generator methode can handel separate one image but sofar i can see evaluate_generator methode wants a Image and Lable. This is not what i need and it is not always a Lable there.

In [7]:     model.evaluate_generator(validation_generator, steps=validation_steps, verbose=1)

**Comment by ZohebAbai** [comment]
Hi,

Can you try rescaling the test images to 1/255.0 before predicting? I think there's the problem. Use the validation imagedatagenerator function. Let us know if it solves the problem.

Side note: In notebook 'CATSvsDOGS_2000_1000_inceptionV3_softmax-Copy5_sent.ipynb' there is validation_data=test_generator inside model.fit_generator, but there is no test_generator defined. As your code ran successfully so I assume you might have changed the name 'test_generator' to 'validation_generator' in your 1st block of code later after the code run.

**Comment by ethex2020** [comment]
Hello,

thank you for the hint normalization!
preprocess_input() should do normalization?
No method preprocess_input() and normalization, it works! 
 

```
# load and pre-process data
def preprocess_input(x):
       return x.astype('float32').reshape((-1,) + input_shape) / 255
```
 
https://keras.io/examples/mnist_net2net/





What i  don't understand is, why the example works on one image:


```
model = ResNet50(weights='imagenet')

img_path = 'elephant.jpg'
img = image.load_img(img_path, target_size=(224, 224))
x = image.img_to_array(img)
x = np.expand_dims(x, axis=0)
x = preprocess_input(x)

preds = model.predict(x)
```
https://keras.io/api/applications/

but in a loop where separately one image is given in the process, it won‘t work?


#


Thank you for the hint `validation_data=test_generator`!
You guess right!


Name convention confusion ???

https://keras.io/examples/mnist_net2net/

(x_train, y_train), (**x_test**, **y_test**) = mnist.load_data()

but 

acc: 0.9960 - **val_loss:** 0.0011 - **val_acc**: 0.9925

it should be renamed to

acc: 0.9960 -  **test_loss:** 0.0011 - **test_acc**: 0.9925

or

(x_train, y_train), (**x_val**, **y_val**) = mnist.load_data()

And the documentation for` monitor=" "` should be optimized:
`callbacks=[ModelCheckpoint(   , monitor="val_acc")]`

**Comment by jvishnuvardhan** [comment]
@ethex2020 Is this still an issue for you? 

Can you please share a simple standalone code to reproduce the issue? Also, test with recent TF version and let us know whether the issue persists with recent versions. Thanks!
